### PR TITLE
Package modem-manager-gui

### PR DESCRIPTION
**Pull request by alesya-h**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  ];
+
+  preConfigure = ''
+    sed -i 's|! test -e "/usr/include/gdbm/gdbm.h"|false|' configure
+    sed -i "s|/etc/NetworkManager/dispatcher.d|$out/etc/NetworkManager/dispatcher.d|" src/scripts/Makefile
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://linuxonly.ru/page/modem-manager-gui;
+    description = "A simple GTK+ based GUI compatible with Modem manager, Wader and oFono system services able to control EDGE/3G/4G broadband modem specific functions. You can check balance of your SIM card, send or receive SMS messages, control mobile traffic consumption and more.";
```
Comment (Synthetica9) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: You might want to wrap this using `''`

```suggestion
    description = ''
      A simple GTK+ based GUI compatible with Modem manager, Wader and oFono
      system services able to control EDGE/3G/4G broadband modem specific
      functions. You can check balance of your SIM card, send or receive SMS
      messages, control mobile traffic consumption and more.
    '';
```

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  src = fetchhg {
+    url = https://linuxonly@bitbucket.org/linuxonly/modem-manager-gui;
+    rev = "8bcf842";
+    sha256 = "04p08gzvi068bd3slbnaghajwzp4h5jacib640k5shngqimkl84p";
+  };
+
+  buildInputs = [
+    pkgconfig
+    gtk3-x11
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: Why not just `gtk3`?

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  rev = "v${version}";
+
+  src = fetchhg {
+    url = https://linuxonly@bitbucket.org/linuxonly/modem-manager-gui;
+    rev = "8bcf842";
+    sha256 = "04p08gzvi068bd3slbnaghajwzp4h5jacib640k5shngqimkl84p";
+  };
+
+  buildInputs = [
+    pkgconfig
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: Programs executed during build should go to `nativeBuildInputs`.

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    gtkspell3
+
+    python3
+    perldevelPackages.Po4a
+    itstool
+    libappindicator-gtk3
+  ];
+
+  preConfigure = ''
+    sed -i 's|! test -e "/usr/include/gdbm/gdbm.h"|false|' configure
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: Could you create a patch file instead? `sed` calls get outdated pretty fast without any warning.

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  version = "0.4.17";
+  rev = "v${version}";
+
+  src = fetchhg {
+    url = https://linuxonly@bitbucket.org/linuxonly/modem-manager-gui;
+    rev = "8bcf842";
+    sha256 = "04p08gzvi068bd3slbnaghajwzp4h5jacib640k5shngqimkl84p";
+  };
+
+  buildInputs = [
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: It seems you forgot to add meson to nativeBuildInputs.

Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: Even though meson and ninja are specified on the project page, it also builds with just ./configure and make. I tried to make it work with meson, but for some reason it turned out to be much harder.

File: pkgs/applications/networking/modem-manager-gui/default.nix (hunk 2 above)
Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: I didn't knew it existed - `nix-env -f "<nixpkgs>" -qaP|grep gtk3` doesn't list gtk3, only gtk3-x11. Is there a command that would have listed gtk3?

Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: fixed

File: pkgs/applications/networking/modem-manager-gui/default.nix (hunk 3 above)
Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: fixed

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    perldevelPackages.Po4a
+  ];
+
+  buildInputs = [
+    gtk3
+    glib
+    gdbm
+    gtkspell3
+
+    itstool
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: `itstool` belongs to `nativeBuildInputs`

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  src = fetchhg {
+    url = https://linuxonly@bitbucket.org/linuxonly/modem-manager-gui;
+    rev = "8bcf842";
+    sha256 = "04p08gzvi068bd3slbnaghajwzp4h5jacib640k5shngqimkl84p";
+  };
+
+  patches = [ ./remove-fhs-gdbm-check.patch ];
+
+  nativeBuildInputs = [
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: Meson is the preferred build system. Could you use that instead of autotools?

File: pkgs/applications/networking/modem-manager-gui/default.nix (hunk 5 above)
Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]:
```
...
Build targets in project: 24
Found ninja-1.8.2 at /nix/store/35wvn5g0igx8xpqzm66j3anpi0icy6rg-ninja-1.8.2/bin/ninja
meson: enabled parallel building
building
build flags: -j1 -l1
[1/66] Generating modem-manager-gui.appdata.xml_merge with a custom command.
[2/66] Generating modem-manager-gui.desktop_merge with a custom command.
[3/66] Generating man-de with a custom command.
FAILED: man/de/modem-manager-gui.1.gz
/build/hg-archive/man/manhelper.py ../man/de/de.po man/de/modem-manager-gui.1.gz
/bin/sh: /build/hg-archive/man/manhelper.py: not found
ninja: build stopped: subcommand failed.
builder for '/nix/store/jnn4w26z9z13kqgij31s0mw0y1p8p2m0-modem-manager-gui.drv' failed with exit code 1
```

Comment (jtojnar) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: It cannot find the `/usr/bin/env` in the shebang of the script You will need to do

```
postPatch = ''
  patchShebangs man/manhelper.py
'';
```

possibly preceded by `chmod +x`ing the file.

Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: manhelper.py is already executable, so it worked without chmod, but then the script itself failed:

```
meson: enabled parallel building
building
build flags: -j1 -l1
[1/66] Generating modem-manager-gui.appdata.xml_merge with a custom command.
[2/66] Generating modem-manager-gui.desktop_merge with a custom command.
[3/66] Generating man-de with a custom command.
FAILED: man/de/modem-manager-gui.1.gz
/build/hg-archive/man/manhelper.py ../man/de/de.po man/de/modem-manager-gui.1.gz
Traceback (most recent call last):
  File "/build/hg-archive/man/manhelper.py", line 20, in <module>
    for line in intfile:
  File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 363: ordinal not in range(128)
ninja: build stopped: subcommand failed.
builder for '/nix/store/v9v1sbnpr6xazzp8qrr98yb32wnpmyi9-modem-manager-gui.drv' failed with exit code 1
```

File: pkgs/applications/networking/modem-manager-gui/default.nix (hunk 6 above)
Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: fixed

File: pkgs/applications/networking/modem-manager-gui/default.nix (hunk 5 above)
Comment (jtojnar) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: It is one of issues with our python. You can fix it by adding `LC_ALL = "en_US.utf-8"` attribute to the derivation and `glibcLocales` to `nativeBuildInputs`.

Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/default.nix]: That worked. Updated the package to build with meson and ninja.

File: pkgs/applications/networking/modem-manager-gui/remove-fhs-gdbm-check.patch [hunk 8]
```diff
@@ -0,0 +1,16 @@
+--- a/configure	2018-10-25 14:19:33.361529316 +1100
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/remove-fhs-gdbm-check.patch]: This should no longer be required with meson.

Comment (alesguzik) [reply pkgs/applications/networking/modem-manager-gui/remove-fhs-gdbm-check.patch]: fixed

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 9]
```diff
@@ -0,0 +1,50 @@
+{ stdenv, buildEnv, pkgconfig, python3, fetchhg, gtk3, glib, gdbm, gtkspell3, itstool, libappindicator-gtk3, perldevelPackages, glibcLocales, meson, ninja }:
+
+stdenv.mkDerivation rec {
+  name = "modem-manager-gui";
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: `name` is supposed to be `modem-manager-gui-${version}`. At least until https://github.com/NixOS/rfcs/pull/35 is implemented.

File: pkgs/applications/networking/modem-manager-gui/default.nix [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  version = "0.4.17";
+  rev = "v${version}";
```
Comment (jtojnar) [inline pkgs/applications/networking/modem-manager-gui/default.nix]: Why is this here?

**Comment by alesguzik** [comment]
@jtojnar thanks. ModemManager backend works for my hardware, so maybe another time.

File: pkgs/top-level/perl-packages.nix [hunk 11]
```diff
@@ -12389,7 +12389,7 @@ let
       export PERL_MB_OPT="--install_base=$out --prefix=$out"
       substituteInPlace Po4aBuilder.pm --replace "\$self->install_sets(\$self->installdirs)->{'bindoc'}" "'$out/share/man/man1'"
     '';
-    buildPhase = "perl Build.PL --install_base=$out; ./Build build";
+    buildPhase = "perl Build.PL --install_base=$out --install_path=\"sbin=$out/bin\" --install_path=\"lib=$out/lib/perl5/site_perl\"; ./Build build ";
```
Comment (jtojnar) [inline pkgs/top-level/perl-packages.nix]: Why a this needed?

Is not there something like `python.sitePackages` for Perl so the path does not need to be specified manually?

Comment (alesguzik) [reply pkgs/top-level/perl-packages.nix]: I don't know what `python.sitePackages` is. Without this change the package was broken - whenever you try to run it's executables the app can't find it's own packages. I copied this buildPhase from another perl package and it fixed that issue.

Comment (jtojnar) [reply pkgs/top-level/perl-packages.nix]: Try `perl.libPrefix`, see https://github.com/NixOS/nixpkgs/blob/b547594fa67cd667678b628093ca031b941a3457/pkgs/development/interpreters/perl/default.nix#L105

Comment (alesguzik) [reply pkgs/top-level/perl-packages.nix]: fixed

**Comment by jtojnar** [comment]
Thank you.
